packages/ui/src/containers/PhoneForm/use-sms-sign-in.ts

```diff
@@ -0,0 +1,89 @@
+import type { SignIn } from '@logto/schemas';
+import { SignInIdentifier } from '@logto/schemas';
+import { useState, useMemo, useCallback } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+import { sendSignInSmsPasscode } from '@/apis/sign-in';
+import type { ErrorHandlers } from '@/hooks/use-api';
+import useApi from '@/hooks/use-api';
+import type { ArrayElement } from '@/types';
+import { UserFlow } from '@/types';
+
+export type MethodProps = ArrayElement<SignIn['methods']>;
+
+const useEmailSignIn = ({ password, isPasswordPrimary, verificationCode }: MethodProps) => {
+  const [errorMessage, setErrorMessage] = useState<string>();
+  const navigate = useNavigate();
+
+  const errorHandlers: ErrorHandlers = useMemo(
+    () => ({
+      'guard.invalid_input': () => {
+        setErrorMessage('invalid_phone');
+      },
+    }),
+    [setErrorMessage]
+  );
+
+  const clearErrorMessage = useCallback(() => {
+    setErrorMessage('');
+  }, []);
```

Comment:
This is just a pure setState so we don't have to wrap it in a `useCallback`

Reply:
updated

Reply:
It's more like a safe guarantee of single-ref for the returned function reference. As it will be used as a dependency in other useEffect hooks. Even though setErrorMessage is a pure setState method, also have to make sure the returned wrapper method reference won't change on each hook execution cycle.